src/vos/vos_dtx.c

```diff
@@ -867,45 +836,39 @@ vos_dtx_commit_one(struct vos_container *cont, struct dtx_id *dti,
 
 		DCE_XID(dce) = *dti;
 		DCE_EPOCH(dce) = epoch;
-		if (dth->dth_oid_array != NULL) {
-			if (dth->dth_oid_cnt == 1) {
-				DCE_OID(dce) = dth->dth_oid_array[0];
-				/* The dec_dkey_hash is the real dkey hash. */
-				DCE_DKEY_HASH(dce) = dth->dth_dkey_hash;
-			} else {
-				struct umem_instance	*umm;
-				size_t			 size;
-				umem_off_t		 rec_off;
-
-				umm = vos_cont2umm(cont);
-				size = sizeof(daos_unit_oid_t) *
-					dth->dth_oid_cnt;
-				rec_off = umem_zalloc(umm, size);
-
-				if (umoff_is_null(rec_off)) {
-					D_ERROR("No space to store CMT DTX OID "
-						DF_DTI"\n", DP_DTI(dti));
-					*fatal = true;
-					D_GOTO(out, rc = -DER_NOSPACE);
-				}
+		DCE_OID(dce) = dth->dth_leader_oid;
+		DCE_DKEY_HASH(dce) = dth->dth_dkey_hash;
 
-				memcpy(umem_off2ptr(umm, rec_off),
-				       dth->dth_oid_array, size);
-				DCE_OID_OFF(dce) = rec_off;
-				/* Reuse dec_dkey_hash as the OID count. */
-				DCE_OID_CNT(dce) = dth->dth_oid_cnt;
-			}
-		} else {
-			D_ASSERT(dth->dth_oid_cnt == 0);
+		if (dth->dth_oid_array == NULL) {
+			dce->dce_oids = &DCE_OID(dce);
+			dce->dce_oid_cnt = 1;
+			goto insert;
+		}
 
-			DCE_OID(dce) = dth->dth_leader_oid;
-			/* The dec_dkey_hash is the real dkey hash. */
-			DCE_DKEY_HASH(dce) = dth->dth_dkey_hash;
+		if (dth->dth_oid_cnt == 1) {
+			dce->dce_oid_inline = dth->dth_oid_array[0];
+			dce->dce_oids = &dce->dce_oid_inline;
+			dce->dce_oid_cnt = 1;
+			goto insert;
 		}
-	}
 
-	dce->dce_reindex = 0;
+		D_ASSERT(dth->dth_oid_cnt > 1);
+
+		size = sizeof(daos_unit_oid_t) * dth->dth_oid_cnt;
+		D_ALLOC(dce->dce_oids, size);
```

Comment:
minor, can use D_ALLOC_ARRAY

Reply:
The "size" will be used for both the allocation and the subsequent memcpy.

Reply:
minor, I think it would be better then to use sizeof(*dce->dce_oids) rather than the type.   Technically, the compiler would likely optimize out the duplication

Reply:
Yes, originally, the dce_oids seemed defined as a union. I forgot to change it after redefining it.